### PR TITLE
[breaking] Fix board attach CLI command, and remove gRPC API

### LINKED ISSUE #1925: Rework board attach command
The `board attach` command has a broken implementation:

## Inconsistent port argument format

Currently, the command will only accept a port when passed in as a full path (e.g.: `serial:///dev/cu.usbmodem14301`). This is inconsistent with all the other commands that accept a port argument, which allow the more user friendly address (e.g.: `/dev/cu.usbmodem14301`) to be used.

The command must accept the same address format as the other commands.

In order to allow a protocol of the port to be specified in addition to the address in cases where this is necessary, the command should also support a `--protocol` flag just as in the other commands that accept a port argument. This information will be stored in the `cpu.protocol` field of `sketch.json`.

Support for the previous port format should be retained.

## Outdated board identification feature

Currently, when a port is passed to the command, it attempts to automatically identify the board on that port and save its FQBN and name in the `cpu.fqbn` and `cpu.name` fields of `sketch.json`.

```
$ arduino-cli board attach serial:///dev/cu.usbmodem614601
Board found: Arduino MKR1000...
Selected fqbn: arduino:samd:mkr1000

$ cat sketch.json
{
  "cpu": {
  "fqbn": "arduino:samd:mkr1000",
  "name": "Arduino MKR1000",
  "port": "serial:///dev/cu.usbmodem614601"
  }
}
```

This feature uses the obsolete [`github.com/arduino/board-discovery` module](https://github.com/arduino/board-discovery).

Although the feature could be modernized, its value is questionable. The available resources would be better allocated to more important work than to updating and maintaining the feature. For this reason, the auto-detection of board data from port feature should simply be removed.

### PR DESCRIPTION
**Pull request by silvanocerza**
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] `UPGRADING.md` has been updated with a migration guide (for breaking changes)

## What kind of change does this PR introduce?

Remove an unused neglected broken feature.

## What is the current behavior?

`board attach` CLI command and `BoardAttach` gRPC interface command are both broken and neglected. Their behaviour is unclear and buggy.

## What is the new behavior?

Remove `board attach` CLI command and `BoardAttach` gRPC interface command and all their usages from the project.

## Does this PR introduce a breaking change, and is [titled accordingly](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#breaking)?

Yes and yes.

## Other information

Closes #1925.

### CODE REVIEW

File: cli/board/board.go [hunk 1]
```diff
@@ -31,12 +31,9 @@ func NewCommand() *cobra.Command {
 		Short: tr("Arduino board commands."),
 		Long:  tr("Arduino board commands."),
 		Example: "  # " + tr("Lists all connected boards.") + "\n" +
-			"  " + os.Args[0] + " board list\n\n" +
-			"  # " + tr("Attaches a sketch to a board.") + "\n" +
-			"  " + os.Args[0] + " board attach serial:///dev/ttyACM0 mySketch",
+			"  " + os.Args[0] + " board list\n\n",
```
Comment (cmaglie) [inline cli/board/board.go:34]:
```suggestion
			"  " + os.Args[0] + " board list",
```

File: docs/UPGRADING.md [hunk 2]
```diff
@@ -2,6 +2,15 @@
 
 Here you can find a list of migration guides to handle breaking changes between releases of the CLI.
 
+## Unreleased
+
+### `board attach` CLI command and `BoardAttach` gRPC interface command removal
+
+Both the `board attach` CLI command and the `cc.arduino.cli.commands.v1.BoardAttach` gRPC command have been removed.
+
+They are not replaced by any new command, the only way to get the same behaviour is to edit the `sketch.json` file either manually or automatically with other tools.
```
Comment (cmaglie) [inline docs/UPGRADING.md]: Is it true? is `sketch.json` supported in any way after removing `board attach`?

**Review by cmaglie — DISMISSED** [review]
* There is this test to [remove](https://github.com/arduino/arduino-cli/actions/runs/3281876179/jobs/5404514043#step:6:6986):
```
    board_test.go:452: 
        	Error Trace:	/home/runner/work/arduino-cli/arduino-cli/internal/integrationtest/board/board_test.go:452
        	Error:      	Received unexpected error:
        	            	exit status 1
        	Test:       	TestBoardAttachWithoutSketchJson
--- FAIL: TestBoardAttachWithoutSketchJson (2.50s)
```

* you must run a `go mod tidy` on all sub-projects [to fix all the internal go.mod/go.sum](https://github.com/arduino/arduino-cli/actions/runs/3281876181/jobs/5404514253)
* thinking about go.mod, we should also be able to get rid of this dependency: `github.com/arduino/board-discovery` so running a `go mod tidy` on the main repo should show it.

**Comment by cmaglie** [comment]
Ah you already did it most of it, LOL. I should reload the page before starting a review :-)

**Comment by silvanocerza** [comment]
> Ah you already did it most of it, LOL. I should reload the page before starting a review :-)

I just pushed it really, you would have missed it in any case. :sweat_smile: 

By the way not sure about the `sketch.json`, the Sketch specifications docs still mention it and I didn't check other commands to see if it's still in use.

Regarding `go mod tidy` I found out that `commands/daemon/term_example` has a `go.sum` but no `go.mod`, so running it there fixes the root one. 
It's also checked in CI `check-outdated` job, see [this](https://github.com/arduino/arduino-cli/blob/master/.github/workflows/check-go-task.yml#L103), but that's like checking root two times.

**Comment by cmaglie** [comment]
> Regarding go mod tidy I found out that commands/daemon/term_example has a go.sum but no go.mod, so running it there fixes the root one.

I see, I think I forgot to commit the `go.mod` too, BTW since the example do not use any external library I think we can remove the extra `go.sum` and the check from the workflow.

**Comment by silvanocerza** [comment]
I see dependencies checks are failing but I can't manage to run `licensed` locally. Would be great if someone could do it. <3

**Comment by per1234** [comment]
Here you go Silvano: https://github.com/silvanocerza/arduino-cli/pull/5

**Comment by silvanocerza** [comment]
Thanks @per1234, greatly appreciated. :)

**Comment by umbynos** [comment]
Hello @silvanocerza, thanks a lot for this PR! Today with @per1234 @ubidefeo @cmaglie we spoke about this. In the end we decided to not completely deprecate the command, mainly because the users are using this and overall it only needs some love.
To summarize what we discussed:
- drop the current implementation of the board attach cli command (done already in this PR)
- the cli will still use the same `sketch.json` file for saving metadata
- the `board attach` command now has to accept 3 flags `--fqbn`, `--port`, and optionally `--protocol` (regarding this there should already be all the helper functions in [cli/arguments](https://github.com/arduino/arduino-cli/tree/master/cli/arguments).

**Comment by silvanocerza** [comment]
Sorry, no time to completely rework the command like that.
Feel free to keep working on this PR though.

**Comment by cmaglie** [comment]
I've pushed the implementation, but now I have my own review pending... I can approve the my changes by myself but doesn't look like a good plan mmmmmhhhh...

File: cli/arguments/port.go [hunk 3]
```diff
@@ -57,11 +56,12 @@ func (p *Port) AddToCommand(cmd *cobra.Command) {
 }
 
 // GetPortAddressAndProtocol returns only the port address and the port protocol
-// without any other port metadata obtained from the discoveries. This method allows
-// to bypass the discoveries unless the protocol is not specified: in this
-// case the discoveries are needed to autodetect the protocol.
+// without any other port metadata obtained from the discoveries.
+// This method allows will bypass the discoveries if:
```
Comment (per1234) [inline cli/arguments/port.go:60]:
```suggestion
// This method will bypass the discoveries if:
```

The previous sentence did not make sense. My interpretation was that the removed "allows" word was a leftover from the previous wording.

The alternative would be:

```go
// This method allows bypassing the discoveries if:
```

But I think "will" fits better in this context.

File: docs/UPGRADING.md [hunk 4]
```diff
@@ -54,6 +54,16 @@ func InitWithProfile(instance *rpc.Instance, profileName string, sketchPath *pat
 
 The errors are automatically sent to output via `feedback` package, as for the other `Init*` functions.
 
+### `board attach` CLI command changed behaviour
+
+The `board attach` CLI command has changed behaviour: now it just pick whatever port and FQBN is passed as parameter and
+saves it in the `sketch.json` file, without any validity check or board autodetection.
```
Comment (per1234) [inline docs/UPGRADING.md:60]:
```suggestion
The `board attach` CLI command has changed behaviour: now it just picks whatever port and FQBN is passed as parameter
and saves it in the `sketch.json` file, without any validity check or board autodetection.
```

Minor grammar fix.

File: docs/UPGRADING.md [hunk 5]
```diff
@@ -54,6 +54,16 @@ func InitWithProfile(instance *rpc.Instance, profileName string, sketchPath *pat
...
+
+### `cc.arduino.cli.commands.v1.BoardAttach` gRPC interface command removal
+
+The `cc.arduino.cli.commands.v1.BoardAttach` gRPC command have been removed. This feature is no more available throguh
```
Comment (per1234) [inline docs/UPGRADING.md:64]:
```suggestion
The `cc.arduino.cli.commands.v1.BoardAttach` gRPC command has been removed. This feature is no longer available through
```

Typo and minor grammar fixes.

File: cli/board/attach.go [hunk 6]
```diff
@@ -16,69 +16,60 @@
...
 
-var (
-	port arguments.Port
-)
-
 func initAttachCommand() *cobra.Command {
+	var port arguments.Port
 	attachCommand := &cobra.Command{
-		Use:   fmt.Sprintf("attach -p <%s>|-b <%s> [%s]", tr("port"), tr("FQBN"), tr("sketchPath")),
+		Use:   fmt.Sprintf("attach [-p <%s>] [-b <%s>] [%s]", tr("port"), tr("FQBN"), tr("sketchPath")),
```
Comment (per1234) [inline cli/board/attach.go:31]:
```suggestion
		Use:   fmt.Sprintf("attach [-p <%s> [-l <%s>]] [-b <%s>] [%s]", tr("port"), tr("protocol"), tr("FQBN"), tr("sketchPath")),
```

It seems to me this effectively communicates the nature of the `-l` flag as a supplement to the `-p` flag, which may only be used in combination with that flag.

**Review by per1234 — CHANGES_REQUESTED** [review]

File: cli/board/attach.go (hunk 6 above)
Comment (cmaglie) [reply cli/board/attach.go:31]: Nice!

**Review by per1234 — APPROVED** [review]

**Comment by umbynos** [comment]
Is it expected that running `arduino-cli board attach Blink/` an empty `sketch.json` is created?
```
{
  "cpu": {
    "fqbn": ""
  }
}
```
I also would implement some test to catch this weird behavior.

The `--format json` is not working correctly

**Comment by cmaglie** [comment]
I've changed the PR so now:
* it implements the `board attach` command by saving the user's preferences in the `sketch.yaml`
* the CLI now completely ignores `sketch.json`
* fix the json and text output of the command

IMHO this is the best compromise because:
* We still provide the `board attach` functionality, but the only visible change to the users is the fact that we use `sketch.yaml` instead of `sketch.json`.
* The Web Editor can continue to use the `sketch.json` without overlapping with the Arduino CLI.
* The Arduino CLI will use his own project file `sketch.yaml` that is well-defined and documented.

**Comment by ubidefeo** [comment]
Looks good to me.
I'd review how the default `protocol` is created when adding the port parameter, maybe it's not necessary if empty

**Comment by cmaglie** [comment]
> I'd review how the default protocol is created when adding the port parameter, maybe it's not necessary if empty

I've pushed an update to fix this problem.

File: docs/sketch-project-file.md [hunk 7]
```diff
@@ -115,3 +115,24 @@ When using the profile-based build, the globally installed platforms and librari
...
 and libraries installed in the system.
+
+## Default flags for Arduino CLI usage
+
+The sketch project file may be used to set the default value for some command line flags of the Arduino CLI, in
+particular:
+
+- The `default_fqbn` directive sets the default value for the `--fqbn` flag
+- The `default_port` directive sets the default value for the `--port` flag
+- The `default_protocol` directive sets the default value for the `--protocol` flag
```
Comment (per1234) [inline docs/sketch-project-file.md:126]:
```suggestion
- The `default_fqbn` key sets the default value for the `--fqbn` flag
- The `default_port` key  sets the default value for the `--port` flag
- The `default_protocol` key sets the default value for the `--protocol` flag
```

I think "key" is the correct terminology here.

References:

- https://yaml.org/spec/1.2.2/ext/glossary/#:~:text=M-,Mapping,-A%20kind%20of
- https://www.yaml.info/learn/index.html#:~:text=items%20and%20more.-,Mapping,-The%20most%20common

File: docs/sketch-specification.md [hunk 8]
```diff
@@ -68,14 +68,12 @@ the `data` folder, so any non-code files outside the `data` folder are stripped.
 
 #### `sketch.json`
 
-Arduino CLI and Arduino Web Editor use a file named sketch.json, located in the sketch root folder, to store sketch
-metadata.
+Arduino Web Editor use a file named `sketch.json`, located in the sketch root folder, to store sketch metadata. This
+file is not used by the Arduino CLI or the Arduino IDE, if you're not an Arduino Web Editor user you can safely ignore
+it.
```
Comment (per1234) [inline docs/sketch-specification.md:73]:
```suggestion
Arduino Web Editor uses a file named `sketch.json`, located in the sketch root folder, to store sketch metadata. This
file is not used by Arduino CLI or Arduino IDE, if you're not an Arduino Web Editor user you can safely ignore the file.
```

- Grammar fix
- Remove some occurrences of "the" (they seemed superfluous to me)
- Make it clear what "it" was referring to (it seemed a little ambiguous to me)

**Review by per1234 — CHANGES_REQUESTED** [review]

File: docs/UPGRADING.md (hunk 4 above)
Comment (umbynos) [inline docs/UPGRADING.md]: needs fixing

File: cli/arguments/fqbn.go [hunk 9]
```diff
@@ -69,17 +69,15 @@ func (f *Fqbn) Set(fqbn string) {
 // - it tries to autodetect the board connected to the given port flags
```
Comment (per1234) [inline cli/arguments/fqbn.go:69]: Please update the reference to `sketch.json` in the comment at line 68 (line is out of the range allowed by PR review suggestions system):

https://github.com/arduino/arduino-cli/blob/ca3929706b537a871dd3804421161a0175dac921/cli/arguments/fqbn.go#L68

**Review by per1234 — CHANGES_REQUESTED** [review]

File: docs/UPGRADING.md (hunk 4 above)
Comment (cmaglie) [reply docs/UPGRADING.md]: I changed it in the next commits.

File: cli/arguments/fqbn.go (hunk 9 above)
Comment (cmaglie) [reply cli/arguments/fqbn.go:69]: There was also another instance of `sketch.json` in the sketch specification: https://github.com/arduino/arduino-cli/pull/1930/commits/eecd894e635d8c59cce228b5f0dcacbf5d637a7e

**Review by umbynos — APPROVED** [review]

**Review by per1234 — APPROVED** [review]
The line removal code in `arduino/sketch/yaml.go` will not handle some forms of valid YAML:

https://github.com/arduino/arduino-cli/blob/eecd894e635d8c59cce228b5f0dcacbf5d637a7e/arduino/sketch/yaml.go#L53-L63

However, it does handle the form of YAML generated by Arduino CLI correctly and I think that the great majority of users will never modify this part of the **sketch project file** via any other mechanism. So I don't think it is worth worrying about at this time.

Other than that, everything is working perfectly for me and the documentation looks great.

Thanks Cristian!

**Comment by cmaglie** [comment]
> The line removal code in arduino/sketch/yaml.go will not handle some forms of valid YAML:

Yes, the YAML format is quite complex, handling all possible variations means writing our own parser (and I read the specification... there are 72631 ways to write a string in YAML).

Also, the existing libraries don't seem capable of changing only a key/value pair without touching the rest of the file, so for now this is the best compromise I found.

**Pull request merged**